_posts/2021-08-23-using-lra.adoc

```diff
@@ -0,0 +1,490 @@
+[2] 2434984
+----
+
+Start an LRA at the first service:
+
+[source,bash]
+----
+$ LRA_URL=$(curl http://localhost:8080/hello/start)
+hello with context http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_34
+----
+
+Kill the first service
+
+[source,bash]
+----
+$ kill 2434936
+2021-08-11 16:02:24,542 INFO  [io.quarkus] (Shutdown thread) narayana-lra-quickstart stopped in 0.003s
+----
+
+Now, with only the second service running, try ending the LRA:
+
+[source,bash]
+----
+$ curl --header "Long-Running-Action: $LRA_URL" http://localhost:8082/hello
+hello with context http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_34
+completing http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_34
+Hello RESTEasy
+----
+
+The LRA will still be running, which you can verify by querying the coordinator: `curl http://localhost:50000/lra-coordinator`.
+
+To finish the LRA restart the failed service (which was listening on port 8080):
+
+[source,bash]
+----
+$ ./target/narayana-lra-quickstart-1.0.0-SNAPSHOT-runner -Dquarkus.lra.coordinator-url=http://localhost:50000/lra-coordinator &
+[3] 2435130
+----
+
+Recovery processing is periodic (the default period between recovery passes is 2 minutes). If you cannot wait 2 minutes then you may manually trigger a recovery cycle via the coordinators recovery endpoint: 
+
+[source,bash]
+----
+$ curl http://localhost:50000/lra-coordinator/recovery
+completing http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_34
+[]
+----
+
+The item to note here is that the restarted service received the completion notification (`completing ...`).
+The result of the request to run a recovery cycle is a json array of recovering LRAs (in this example the list is empty because the last LRA has now finished as indicated by the empty json array `[]`).
+
+Clean up by stopping the two services (`kill 2434936 2434984`).
+
+== Appendix 1
+
+=== Embedded Coordinators
+
+Since a coordinator is just a JAX-RS application they can be embedded with JAX-RS services quite easily, just add the LRA coordinator dependency to your pom.xml file:
+
+[source,xml]
+----
+    <dependency>
+      <groupId>org.jboss.narayana.rts</groupId>
+      <artifactId>lra-coordinator-jar</artifactId>
+      <version>5.12.0.Final</version>
+    </dependency>
+----
+
+and since, by default, quarkus only allows one application per deployment you will need to add the the following property to the application config file (`src/main/resources/application.properties`):
+
+[source,xml]
+----
+quarkus.resteasy.ignore-application-classes=true 
+----
+
+The same caveats as described in the <<_lra-coordinators,the section on coordinators>> still apply:
+
+- no support for native executables.
+- each instance requires dedicated storage for its transaction logs (since sharing transaction stores is not supported).
```

Comment:
Do we consider starting two coordinators in the same app or a participant and coordinator.
If the later then the participant does not require to work with the transaction log storage.  And maybe the second caveat is not needed to be mentioned?

Reply:
Two coordinators in one app does not make much sense (and I cannot conceive of a use case).

But a coordinator embedded with a participant does make sense and the associated storage does need to be managed: the fact that a participant is running in the same JVM as the coordinator makes no difference to the fact that the storage needs to be dedicated for use by only that coordinator.

Reply:
I see, my point was that the participant does not need any storage that would be related to Narayana. The object store is used only by coordinator. The participant does not touch it. Participant will usually have own "unrelated" storage.

The note is not problematic just it seemed to me to hint that the participant could be somehow related to object store which should not be so.

Reply:
Your point is noted but embedding a coordinator with the participant implies that the associated storage for the transaction logs cannot be shared. I reworded the sentence to make it clear:

`dedicated storage for its transaction logs` -> `dedicated storage for the coordinators' transaction logs`